### ISSUE TITLE
Decouple configuration management from netdata redistributable binaries

### ISSUE THREAD

**Issue by avishefi** [firehol/netdata#2622]
Hi,
The existing makeself zip file `netdata-v*-<hash>.run` packaged under the binary-packages GitHub repository unzips and creates a user and a group and configures the logrotate file and systemd/initd unit files. Additionally, the existing Ansible Galaxy roles are using the compile-from-source installation method, but contain only partial configuration as some of it is coming from the netdata installer directly.
The compile-from-source installation method adds runtime dependencies to environments which sometimes conflict with security policies that forbid the installation of compilers on production environments where netdata is to be deployed.

To decouple these shared responsibilities we should release netdata as a standalone binary using GitHub releases on the netdata repository, instead of the binary-packages repo. Once we have that, we can then use configuration management tools (e.g. Ansible, Puppet, etc.) to download netdata from GitHub releases, create the user & group, and setup the systemd unit files, log rotation, netdata config files, etc. All configuration details would become part of the configuration management units, and will have complete control over the installation and update of netdata.

This strikes a good balance between the responsibility of netdata and the amount of control that a system operator is given using their own configuration management tools.

**Comment by ktsaou** [comment]
Hi,

1. the `.run` packages have a `--noexec` option. Have you tried it? It is supposed to just unpack the files and exit.

2. the script run by the binary packages will not do anything on the system, if it has already be done. So, if you have already added user and group `netdata`, added a logrotate file, installed a netdata.service file, etc, the script will not overwrite them. Have you tried it?

**Comment by ktsaou** [comment]
I also see that the `.run` packages support ` [--] [options to the embedded script]`. This means we can change this script to perform some of its functions given a few command line switches.

This is the script that runs after unpacking the binary netdata: https://github.com/firehol/netdata/blob/master/makeself/install-or-update.sh

**Comment by avishefi** [comment]
The --noexec is a possible solution, I didn't notice it.

However:
* .run script requires administrative permissions even when unzipping using --noexec
* when running it with `sudo sh *.run` it requires accepting the license, which is a problem when installing netdata automatically using configuration management tools
* when running with `sudo sh *.run --target .` it unpacks other directories such as `var`, `usr`, `include`, `lib` & `bin` which is great when used standalone, but completely overrides the installed curl & libcurl, and potentially confuses which binary/library/headers are used on netdata and on the rest of the system where it is installed.

I think we can easily get around this by not requiring administrative permissions & license approval for an automated installation, combined with a configuration management recipe which will take care of all the rest of the details.

The downside is that there will always be the duplication between makeshell scripts and the configuration management recipes as these will tend to get updated after a netdata release is made available since they will be maintained on other git repos (maybe maintain them on the netdata repo?).

**Comment by ktsaou** [comment]
hm...

1. I can change makeself [to not need admin privileges](https://github.com/ktsaou/netdata/blob/93e8f638ba13390e877ec65af8f94e027e676931/makeself/jobs/99-makeself.install.sh#L72-L72).

2. Add `--accept` to the command line

3. I suggest not to install binary archives merged into the system. This has a lot of issues and should be avoided. Packages in the standard directories should be installed only by the package manager, or things can go really weird (even `./netdata-installer.sh` should install netdata in `/opt` by default, but initially I didn't have this knowledge). Also, the binary package has `/opt/netdata` all over the place. You can relocate it, but it will require quite some configuration.

I think we need to somehow differentiate 3 tasks:

1. prepare/modify the system to allow running netdata. That is, add netdata user and group, install logrotate file, startup script, etc.

2. install netdata to a directory without touching anything else on the system. That is, change permissions on netdata directories and files, so that it will be able to run properly.

3. configure netdata the way you like it. That is, change netdata configuration to suit your needs or change other apps or system configuration to allow netdata do whatever you like it to do.

I think No 2 should never be handled by your provisioning tools (or a lot of duplication of work will be required). This is purely netdata internal work you should not care (why bother changing the permissions of web files, or adding the setuid bit to apps.plugin?)

Regarding No 1, it may or may not be handled by your provisioning tools. But again why bother with the basics if the netdata package can do this perfectly? Of course, your provisioning tool may complement what netdata already does, to install additional python packages, node.js, fping, freeipmi, etc.

I think No 3 is the one you should focus. For the static netdata binary to access your web server log files it needs additional permissions (add user `netdata` to the group your web server is running), to connect to your database server your db may need to allow it, etc.

So, my suggestion is to not replicate with your provisioning tools everything the package does. I suggest to add command line options to the package to prevent it from doing something that bothers your logic (e.g. prevent it from starting netdata at the end of the installation) and complement with your tools what netdata does.

**Comment by avishefi** [comment]
Sounds good, I gave more thought to it and considering that makeshell is the package manager of netdata, replacing rpm/deb in a way, then no. 1+2 are indeed already handled by it.

So we have 2 small changes which I think I will open a different issue for:
1. add an option to no force admin privileges
2. add `--accept` to the command line

Also, I noticed that on the kickstart file at https://github.com/firehol/netdata/blob/master/kickstart-static64.sh there is no option to add command-line parameters to the `*.run` that is executed.  Do you think this should be added?

**Comment by ktsaou** [comment]
1. I can do it (although it will fail to add system users if not run as root)
2. you have to do it (add `--accept` to `.run` command line)
3. yes, I can add options to kickstart64.sh that will be passed through to `.run`

Keep in mind though, `kickstart64.sh` is just a wrapper to download the latest nightly `.run` file and run it. If you want to control which version is installed, you should just download the run the `.run` file you like.

**Comment by avishefi** [comment]
Hi,
Regarding no. 1 - I think this should be optional for --noexec installations, I don't see a reason to change the existing default behavior.

I will perform the 3 changes myself and submit a pull request.

**Comment by avishefi** [comment]
I reviewed the installation & setup code and realized that the installation script is built with the intention of providing everything and there are quite a few details that are spread across its files. My motivation is to make it easier to control the configuration of netdata, but maintaining these configuration files on a version-controlled repository and then overriding them after netdata is installed is also a good solution.

I got along with the existing `--non-interactive` option which works great. I posted an Ansible role which uses the binaries from the binary-packages repository and does not require compiling when installing netdata, here it is: https://github.com/avishefi/ansible-role-netdata
